diffrax/misc/delays.py

```diff
@@ -0,0 +1,99 @@
+from typing import Callable, Tuple, Union
+
+import equinox as eqx
+import jax.lax as lax
+import jax.numpy as jnp
+import jax.tree_util as jtu
+
+from ..custom_types import Array, Bool, Int, PyTree
+
+
+class Delays(eqx.Module):
+    delays: PyTree[Callable]
+    initial_discontinuities: Union[Array, Tuple]
+    max_discontinuities: Int
+    recurrent_checking: Bool
+    rtol: float
+    atol: float
+
+
+def history_extrapolation_implicit(
+    implicit_step, terms, dense_interp, solver, delays, t0, y0_history, state, args
+):
+    from ..integrate import _HistoryVectorField
+    from ..term import VectorFieldWrapper
+
+    def _cond_fun(_val):
+        _, _, _, _, _, pred, step = _val
+        return (implicit_step & pred) | (jnp.invert(implicit_step) & (step == 0))
+
+    def _body_fun(_val):
+        y_prev, _, dense_info, _, _, _, step = _val
+        is_vf_wrapper = lambda x: isinstance(x, VectorFieldWrapper)
+
+        def _apply_history(x):
+            if is_vf_wrapper(x):
+                vector_field = _HistoryVectorField(
+                    x.vector_field,
+                    t0,
+                    state.tprev,
+                    state.tnext,
+                    dense_info,
+                    dense_interp,
+                    solver.interpolation_cls,
+                    y0_history,
+                    delays.delays,
+                )
+                return VectorFieldWrapper(vector_field)
+            else:
+                return x
+
+        terms_ = jtu.tree_map(_apply_history, terms, is_leaf=is_vf_wrapper)
+        (y, y_error, dense_info, solver_state, solver_result) = solver.step(
+            terms_,
+            state.tprev,
+            state.tnext,
+            state.y,
+            args,
+            state.solver_state,
+            state.made_jump,
+        )
+
+        y_error = jtu.tree_map(
+            lambda _y_error: jnp.where(step < 10, _y_error, jnp.inf),
+            y_error,
+        )
+
+        _pred = (jnp.abs(((y - y_prev) / y)) > delays.rtol).any() | (
+            jnp.abs(y - y_prev) > delays.atol
+        ).any()
```

Comment:
I'm not sure if this is a standard form for this check for DDEs specifically? (The numerical literature tends to do all kinds of odd things for termination conditions, there generally isn't much standardisation.)

I think a more-standard-looking termination condition here would be something like `_pred = rms_norm((y - y_prev) / (atol + rtol * y)) < 1`. In particular this is analogous to the way the adaptive step size controller works.

(Although the termination condition for Newton's method is a little different, and more complicated... so that's still inconsistent.)

Side note, we need to support pytree-valued `y` here. This can be done using the `ω` helper. E.g. for my example above: `_pred = rms_norm(((y**ω - y_prev**ω) / (atol + rtol * y**ω)).ω) < 1`.

WDYT? I make no claim that this approach is the best, merely that it looks more familiar to me.

Reply:
I agree that termination conditions can be quick vague in the litterature, for consistencies sake in `diffrax` i'll switch to `rms_norm`,